plugins/radar-android-google-places/src/main/java/org.radarbase.passive.google.places/GooglePlacesManager.kt

```diff
@@ -0,0 +1,464 @@
+import org.radarcns.passive.google.PlacesType
+import org.slf4j.LoggerFactory
+import java.util.concurrent.TimeUnit
+import kotlin.math.pow
+
+class GooglePlacesManager(service: GooglePlacesService, @get: Synchronized private var apiKey: String, private val placeHandler: SafeHandler) : AbstractSourceManager<GooglePlacesService, GooglePlacesState>(service) {
+    private val placesInfoTopic: DataCache<ObservationKey, GooglePlacesInfo> = createCache("android_google_places_info", GooglePlacesInfo())
+
+    private val placesProcessor: OfflineProcessor
+    // Delay in seconds for exponential backoff
+    private val maxDelay: Long = 43200
+    private val baseDelay: Long = 300
+    @get: Synchronized
+    @set: Synchronized
+    private var numOfAttempts: Int = 0
+    private val preferences: SharedPreferences = service.getSharedPreferences(GooglePlacesManager::class.java.name, Context.MODE_PRIVATE)
+
+    private val isPermissionGranted: Boolean
+        get() = checkLocationPermissions()
+    private val placesClientCreated: Boolean
+        get() = state.placesClientCreated.get()
+    private val fromBroadcastRegistration: Boolean
+        get()= state.fromBroadcast.get()
+    var shouldFetchPlaceId: Boolean
+        get() = state.shouldFetchPlaceId.get()
+        set(value) = state.shouldFetchPlaceId.set(value)
+    var shouldFetchAdditionalInfo: Boolean
+        get() = state.shouldFetchAdditionalInfo.get()
+        set(value) = state.shouldFetchAdditionalInfo.set(value)
+    var limitByPlacesCount: Int
+        get() = state.limitByPlacesCount
+        set(value) { state.limitByPlacesCount = value }
+    var limitByPlacesLikelihood: Double
+        get() = state.limitByPlacesLikelihood
+        set(value) { state.limitByPlacesLikelihood = value }
+    private var willDisconnect: Boolean
+        get() = state.isDisconnecting.get()
+        set(value) = state.isDisconnecting.set(value)
+    private val currentPlaceFields: List<Place.Field>
+        get() = if (shouldFetchPlaceId||shouldFetchAdditionalInfo) listOf(Place.Field.TYPES, Place.Field.ID) else listOf(Place.Field.TYPES)
+    private val detailsPlaceFields: List<Place.Field> =  listOf(Place.Field.ADDRESS_COMPONENTS)
+
+    private var placesBroadcastReceiver: BroadcastRegistration? = null
+    private var placesClient: PlacesClient? = null
+        @Synchronized get() = if (placesClientCreated) field else null
+
+    private lateinit var currentPlaceRequest:FindCurrentPlaceRequest
+    private lateinit var detailsPlaceRequest:FetchPlaceRequest
+
+    init {
+        name = service.getString(R.string.googlePlacesDisplayName)
+        placesProcessor = OfflineProcessor(service) {
+            process = listOf(this@GooglePlacesManager::processPlacesData)
+            requestCode = GOOGLE_PLACES_REQUEST_CODE
+            requestName = GOOGLE_PLACES_REQUEST_NAME
+            wake = false
+        }
+    }
+
+    override fun start(acceptableIds: Set<String>) {
+        register()
```

Comment:
here you can check 

```
delay = (baseDelay + 2.0.pow(numOfAttempts)).toLong()
if (currentTime >  timeOfFailedRequest + delay) {
     register()
     .....
} else {
     disconnect()
}
```

Then you don't need the delayed handler for disconnecting